Expose flake directory to `nix fmt` as `PRJ_ROOT` env var

jfly opened the pull request:
This was discussed in https://github.com/NixOS/nix/issues/8034.

I only implemented this for `nix fmt` because it doesn't let you point
at a flake not on your filesystem.


<!--

IMPORTANT

Nix is a non-trivial project, so for your contribution to be successful,
it really is important to follow the contributing guidelines:

https://github.com/NixOS/nix/blob/master/CONTRIBUTING.md

Even if you've contributed to open source before, take a moment to read it,
so you understand the process and the expectations.

- what information to include in commit messages
- proper attribution
- volunteering contributions effectively
- how to get help and our review process.

-->

## Motivation

<!-- Briefly explain what the change is about and why it is desirable. -->

## Context

<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

edolstra commented on src/nix/formatter-run.md:
```diff
@@ -8,6 +8,9 @@ Flags can be forwarded to the formatter by using `--` followed by the flags.
 
 Any arguments will be forwarded to the formatter. Typically these are the files to format.
 
+The environment variable `FLAKE_DIR` will be set to the absolute path to the
```
This should be something like `NIX_FLAKE_DIR` since we want all out environment variables to be prefixed with `NIX_`.

jfly replied:
I've updated this to `NIX_FLAKE_DIR`

edolstra commented on src/nix/formatter.cc:
```diff
@@ -7,6 +7,28 @@
 
 using namespace nix;
 
+//<<< where should this go?
```
This should go into libutil (see `src/libutil/environment-variables.cc` etc.).

edolstra commented on src/nix/formatter.cc:
```diff
@@ -72,6 +94,12 @@ struct CmdFormatterRun : MixFormatter, MixJSON
         auto evalState = getEvalState();
         auto evalStore = getEvalStore();
 
+        auto flakeRef = parseFlakeRef(fetchSettings, ".", std::filesystem::current_path().string());
```
The flakeref should be obtained from the installable returned by `parseInstallable()` below. `flakeRef.input` has a method `getSourcePath()` that returns the physical path of a flake, if any.

jfly replied:
Thanks, @xokdvium! I've updated the code to use the helper method you pointed me to. I'm not sure about this nullptr check you suggested, though? I think it's impossible to [construct a `ref` with a null pointer](https://github.com/NixOS/nix/blob/0db10fc875d3dafe365d157aed7a6d9344092fc2/src/libutil/include/nix/util/ref.hh#L10-L11).

jfly replied:
Great! Marking this thread resolved.

Mic92 commented on src/nix/run.cc:
```diff
@@ -27,14 +28,37 @@ std::string chrootHelperName = "__run_in_chroot";
 
 namespace nix {
 
+/* Convert `env` to a list of strings suitable for `execve`'s `envp` argument. */
+Strings toEnvp(Environment env)
+{
+    Strings envStrs;
+    for (auto & i : env) {
+        envStrs.push_back(i.first + "=" + i.second);
+    }
+
+    return envStrs;
+}
```
I'd say only when we start actually using this in different places.

Ericson2314 commented:
@Mic92 told me about https://github.com/numtide/prj-spec/blob/main/PRJ_SPEC.md I would indeed like to use the `PRJ_SPEC`.

Yes, Nix is getting the directory via flakes, but the formatter shouldn't need to know or care about this. The formatter just wants to know what the "project root" is, now how its caller determined that that location is the project root.

jfly commented:
Sure, that works for me! Conveniently, `treefmt` (the project which led me to this area of the code) already supports PRJ_SPEC, so it honors the `PRJ_ROOT` environment variable.

I've added this change, rebased, and resolved conflicts.

Ericson2314 commented:
Great. Can you fix up the commi messages and PR description?

jfly commented:
@Ericson2314, done!

Mic92 commented on tests/functional/formatter.sh:
```diff
@@ -34,13 +34,38 @@ cat << EOF > flake.nix
+mkdir subflake
+cp ./simple.nix ./simple.builder.sh ./formatter.simple.sh "${config_nix}" "$TEST_HOME/subflake"
+
+cat << EOF > subflake/flake.nix
```
I added some tests for subflakes just in case.

jfly replied:
Thanks, @Mic92!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2025-05-21-nix-team-meeting-minutes-228/65204/1
